docs/proposals/20220707-service-primitive.md

```diff
@@ -0,0 +1,25 @@
+# Service Expose Support
+Authors:
+- [nullday](https://github.com/aseaday)
+
+## Summary
+
+This proposal is designed for exposing service from the application running in the envd container such as Juypter.We introduce a built-in function `expose` for achieving this goal.
+
+## Motivation
+
+There a lot of AI/ML SDK can be accessible from Web or GUI Client. Envd users need to configure their container to use these tools better. For example, a user want to write some code on Juypter in their Chrome.
+
+In the future, we should also keep the possibility for remote or cloud access.
+
+## Goals
+- http/tcp 
+- OS support:
+    - Linux/Mac
+    - wsl access from the windows
+
+## API
+```
+expose(local_port=<LPORT>, host_port=<RPORT>, svc=<SERVICE_NAME>)
```

Comment:
How do we know what command should be run to launch the service in envd?

Reply:
I am not sure wether we should take of this in `expose`. I still think we could't see the envd as a docker which means:
1. A container which has a entrypoint to run as a service
2. If we provide some functions to run something in the near future, It'd better be like `run(juypter)` according the presets we defined for users.
So I think `expose` can be seprated from the application.

Reply:
Then we need a mechanism to run some user-defined commands in the container. `run()` does not work in this scenario since it is just the docker `RUN` command, and is launched during the build process, instead of the runtime. We need to run the exposed services in runtime.

Reply:
> Then we need a mechanism to run some user-defined commands in the container. `run()` does not work in this scenario since it is just the docker `RUN` command, and is launched during the build process, instead of the runtime. We need to run the exposed services in runtime.

Yes. We need a keyword to run some applications like juypter during runtime. So there is going to be another proposal which provide the users to start some applications and maybe `start` like:
```python
def build():
    base(os="ubuntu20.04", language="python3")
    install.python_packages(name = [
        "numpy",
        "juypter"
    ])
    shell("zsh")
    start(app.juypter)
```

Reply:
LGTM.